packages/assets-controllers/src/NftDetectionController.ts

```diff
@@ -602,83 +537,123 @@ export class NftDetectionController extends StaticIntervalPollingController<
-        name,
-        description,
-        attributes,
-        topBid,
-        lastSale,
-        rarityRank,
-        rarityScore,
-        collection,
-      } = nft.token;
-
-      let ignored;
-      /* istanbul ignore else */
-      const { ignoredNfts } = this.#getNftState();
-      if (ignoredNfts.length > 0) {
-        ignored = ignoredNfts.find((c) => {
-          /* istanbul ignore next */
-          return (
-            c.address === toChecksumHexAddress(contract) &&
-            c.tokenId === token_id
-          );
-        });
-      }
-
-      /* istanbul ignore else */
-      if (!ignored) {
-        /* istanbul ignore next */
-        const nftMetadata: NftMetadata = Object.assign(
-          {},
-          { name },
-          description && { description },
-          image_url && { image: image_url },
-          image_thumbnail_url && { imageThumbnail: image_thumbnail_url },
-          image_original_url && { imageOriginal: image_original_url },
-          kind && { standard: kind.toUpperCase() },
-          lastSale && { lastSale },
-          attributes && { attributes },
-          topBid && { topBid },
-          rarityRank && { rarityRank },
-          rarityScore && { rarityScore },
-          collection && { collection },
-        );
+    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+    const updateKey: `${Hex}:${string}` = `${chainId}:${userAddress}`;
+    if (updateKey in this.#inProcessNftFetchingUpdates) {
+      // This prevents redundant updates
+      // This promise is resolved after the in-progress update has finished,
+      // and state has been updated.
+      await this.#inProcessNftFetchingUpdates[updateKey];
+      return;
+    }
+
+    const {
+      promise: inProgressUpdate,
+      resolve: updateSucceeded,
+      reject: updateFailed,
+    } = createDeferredPromise({ suppressUnhandledRejection: true });
+    this.#inProcessNftFetchingUpdates[updateKey] = inProgressUpdate;
 
-        await this.#addNft(contract, token_id, {
-          nftMetadata,
-          userAddress,
-          source: Source.Detected,
-          networkClientId: options?.networkClientId,
+    let next;
+    let apiNfts: TokensResponse[] = [];
+    let resultNftApi: ReservoirResponse;
+    try {
+      do {
+        resultNftApi = await this.#getOwnerNfts(userAddress, chainId, next);
+        apiNfts = resultNftApi.tokens.filter(
+          (elm) =>
+            elm.token.isSpam === false &&
+            (elm.blockaidResult?.result_type
+              ? elm.blockaidResult?.result_type === BlockaidResultType.Benign
+              : true),
+        );
+        const addNftPromises = apiNfts.map(async (nft) => {
+          const {
+            tokenId: TokenId,
+            contract,
+            kind,
+            image: ImageUrl,
+            imageSmall: ImageThumbnailUrl,
+            metadata: { imageOriginal: ImageOriginalUrl } = {},
+            name,
```

Comment:
Since we only use PascalCase for types and classes, could we use a different naming scheme here to prevent confusion?

camelCase would be my choice:

```suggestion
            tokenId,
            contract,
            kind,
            image: imageUrl,
            imageSmall: imageThumbnailUrl,
            metadata: { imageOriginal: imageOriginalUrl } = {},
```